src/common/utils/resolve_reference.py

```diff
@@ -0,0 +1,157 @@
+            items.append(QueryItem(query=f"name={content},isRoot=True"))
+    elif prev_deliminator == "(" and deliminator == ")":  # By query
+        items.append(QueryItem(query=content))
+    elif prev_deliminator == "[" and deliminator == "]":  # By list index
+        items.append(AttributeItem(f"[{content}]"))
+    elif prev_deliminator == ".":  # By attribute
+        items.append(AttributeItem(content))
+    else:  # Fallback to support the old reference format
+        items.append(IdItem(content))
+
+    return reference_to_reference_items(remaining_reference, items, deliminator)
+
+
+@dataclass
+class IdItem:
+    """Points to an id."""
+
+    id: str
+
+    def resolve(self, document: dict | list, data_source: DataSource, get_data_source: Callable) -> dict:
+        # Get the document from the data source
+        return data_source.get(self.id)
+
+
+@dataclass
+class QueryItem:
+    """Query a list for a document."""
+
+    query: str
+
+    def filter(self) -> dict:
+        """Create filter from query string."""
+        query = {}
+        for pair in self.query.split(","):
+            key, value = pair.split("=")
+            query[key] = value
+        return query
+
+    def resolve(self, document: dict | list, data_source: DataSource, get_data_source: Callable) -> dict | None:
+        # We need to search inside the data source for the document
+        if not document:
+            result: list[dict] = data_source.find(self.filter())
+            if not result:
+                raise NotFoundException(
+                    f"No document that match '{self.query}', in data source '{data_source.name}' could be found."
+                )
+            if len(result) > 2:
+                raise ApplicationException(
+                    f"More than 1 document that match '{self.query}' was returned from DataSource. That should not happen..."
+                )
+            return result[0]
+
+        # Search inside an existing document (need to resolve any references first before trying to compare against filter)
+        elements = [
+            resolve_reference(f["ref"], data_source, get_data_source) if is_reference(f) else f for f in document
+        ]  # Resolve any references
+        return next(
+            (f for f in elements if is_same(f, self.filter())), None
+        )  # Find an item that match the given filter
+
+
+@dataclass
+class AttributeItem:
+    """Move into a key of an object or find an item in a list based on index position."""
+
+    path: str
+
+    def resolve(self, document: dict | list, data_source: DataSource, get_data_source: Callable) -> dict:
+        try:
+            result = find(document, [self.path])
+            if is_reference(result):
+                return resolve_reference(result["ref"], data_source, get_data_source)
+            return result
+        except Exception:
+            raise NotFoundException(f"No '{self.path}' inside the '{str(document)}' exist.")
```

Comment:
Could we be more specific in the exceptions we catch here?
Might not need the "try" if we move "NotFoundException" inside "resolve_reference"

Reply:
was a bit hassle to fix the Exception, since the exceptions that the find method raise (index, type, key) are used in several places. so easiest to just catch Exception and throw here

Reply:
So this can be done if needed, but new PR, since don't want to change to much.